### ISSUE TITLE
pasting sql with tabs from vscode breaks things ? 

### ISSUE THREAD

**Issue by Ashtonian** [jakob/Postico#445]
Not sure what the deal is but we use vscode with go which uses go fmt to formats things using tabs. Whenever I copy and paste from postico to vscode and vice-versa they tend to break each other. 

Postico responds with ```ERROR:  type "                    " does not exist```

**Comment by jakob** [comment]
Hi!

I'm not sure what is going on, but it might be related to special whitespace characters.

PostgreSQL is very picky about whitespace. All code points above 0x7F are considered "word" characters. This means that things like non-breaking spaces, unicode line separators, etc. are considered word characters by PostgreSQL.

Previous versions of Postico used to "sanitize" whitespace characters when pasting, but that caused a lot of weird issues, so Postico now "sanitizes" whitespace just before sending things to the server. ("sanitize" in this context means replace special whitespace characters with normal whitespace, eg nbsp -> space). It's possible that something isn't working quite right here.

Could you check which version of Postico you are on?

It would also be awesome if you could send me an unaltered copy of a problematic file? Using the "Save..." button in Postico would be best. (copy-pasting to a different app might not work since other apps also "sanitize" whitespace, especially newlines, on pasting)

If you want to debug yourself, I can warmly recommend the Hex Editor [Hex Fiend](http://ridiculousfish.com/hexfiend/)... it's really great for getting to the bottom of encoding issues.

**Comment by rfink** [comment]
This is continuing to happen in version 1.4.2. I've attached a file saved directly from the editor. The source here was vscode.

[faulty_whitespace.txt](https://github.com/jakob/Postico/files/2383801/faulty_whitespace.txt)

**Comment by jakob** [comment]
@rfink Thanks a lot for the test case! Now I can reproduce it!

Long description: PostgreSQL supports only spaces and tabs for whitespace. All other whitespace characters (eg. non-breaking spaces, unicode line separators, etc) are treated as characters. Your query contains non-breaking spaces.

Since this is a common issue, Postico tries to fix the queries you send to the server by replacing non-beaking whitespaces with normal whitespace. However, I don't want to change any data in string literals, so there is some complicated logic that only replaces non-standard whitespace that is not enclosed in string literals.

Unfortunately there's a bug in the whitespace logic, so any non-standard whitespace before a string literal is not replaced.

I'll see if I can fix the logic.

**Comment by rfink** [comment]
That's awesome, thanks! Love the product, everyone at the office uses it.

**Comment by hwsoderlund** [comment]
I was going to post a bug report about this, but then saw that there is one already. Has there been any progress on this one?

**Comment by hwsoderlund** [comment]
Btw, I am on 1.5.8 now, and this is still happening for me.

**Comment by hwsoderlund** [comment]
FYI, not sure if this is valuable: 

* I copy SQL from a .go file in VS Code
* I paste it into TextEdit
* In TextEdit I copy the text again and paste it back into a plain text file in VS Code

Result: Now the original tabs have been replaced with non breaking spaces.

If I perform the above procedure again, but this time using Sublime instead of TextEdit, the problem does not occur. 

When I paste the text into TextEdit, it looks like it has interpreted the clipboard content as some sort of rich text format, since the pasted text is colorized. This is the exact content I am copying from VS Code:

```sql
	SELECT %s
	FROM users
	WHERE status = $2
	AND name IN (
		'john', 
		'susan'
	);
```

And then after pasting into TextEdit, it looks like this, with the tabs replaced with what I _think_ are non breaking spaces:

![image](https://user-images.githubusercontent.com/3725915/63676648-b0440300-c7eb-11e9-9a60-9c2c257969ec.png)

However, if I use the option `Paste and Match Style` when I paste into TextEdit, the tabs seem to be respected.

So, a solution to this could perhaps be to add a new Paste option to Postico, like `Paste Plain Text` or some such.

**Comment by Julien0002** [comment]
same issue for last vscode update. 
it is very painful

**Comment by aaron9000** [comment]
Same - this is making Postico unusable.

Anyone have a workaround for VS Code users?

**Comment by rfink** [comment]
@aaron9000 I've been using this in the meantime:

https://marketplace.visualstudio.com/items?itemName=bmaupin.copy-without-formatting

**Comment by aaron9000** [comment]
@rfink thank you! This works beautifully.

**Comment by rfink** [comment]
@aaron9000 no problem.

**Comment by hwsoderlund** [comment]
Yes, thanks @rfink for the workaround.

**Comment by Julien0002** [comment]
workaround works for me too.

Many thanks

**Comment by jakob** [comment]
Hi everyone!

Sorry for the slow response to this bug report. I've further investigated the issue, and as far as I can tell, there are two separate issues:

1. When pasting text from VS Code, all spaces are converted to non-breaking spaces

2. Since PostgreSQL does not like non-breaking spaces, Postico tries to replace them with normal spaces (except in string literals), but the code that does this had a bug, and syntax highlighting didn't handle non-breaking spaces correctly either.

We've found a possible workaround for 1., and I've also attempted to finally fix the bugs from 2. I've also added a number of regression tests to make sure this problem doesn't happen again.

You can download the latest build with those fixes here:
https://eggerapps-downloads.s3.eu-west-1.amazonaws.com/postico-5186.zip

We still need to do some further testing, but if everything works out we hope to release an update with these fixes soon.

**Comment by jakob** [comment]
@nicpottier Could you elaborate which issue you are still seeing?

PS: I find TextMate to be really good for debugging whitespace issues (when you turn on "show invisible characters")

**Comment by nicpottier** [comment]
Hi @jakob, apologies for not elaborating. Pasting a query back from Postico into VSCode leaves me with what I'm gathering are non-breaking spaces (codepoint 160, see below). Finally tracked them all down by using `View -> Render Whitespace` which shows normal ASCII spaces as small dots and doesn't treat non breaking spaces any differently. I will probably use a variation of the script below in the future.

Just played with it some more and I'm not able to reproduce it with a brand new query / text I write in either Postico or VSCode, but my original query (which was originally written in VSCode, pasted into Postico, tweaked, then pasted back into VSCode still has some phantom characters) So perhaps the new patched version does indeed address this, just doesn't fix/tweak anything when you paste out.

An example was (of a much longer query): `b.base_language as base_language,`

Just out of curiosity:

```python3
>>> for b in "b.base_language as base_language,":
...   print(b, ord(b))
...
b 98
. 46
b 98
a 97
s 115
e 101
_ 95
l 108
a 97
n 110
g 103
u 117
a 97
g 103
e 101
  160
a 97
s 115
  160
b 98
a 97
s 115
e 101
_ 95
l 108
a 97
n 110
g 103
u 117
a 97
g 103
e 101
, 44
```

So ya, it is non-breaking spaces that are hanging me up.

**Comment by shmcsensei** [comment]
> Hi everyone!
> 
> Sorry for the slow response to this bug report. I've further investigated the issue, and as far as I can tell, there are two separate issues:
> 
> 1. When pasting text from VS Code, all spaces are converted to non-breaking spaces
> 2. Since PostgreSQL does not like non-breaking spaces, Postico tries to replace them with normal spaces (except in string literals), but the code that does this had a bug, and syntax highlighting didn't handle non-breaking spaces correctly either.
> 
> We've found a possible workaround for 1., and I've also attempted to finally fix the bugs from 2. I've also added a number of regression tests to make sure this problem doesn't happen again.
> 
> You can download the latest build with those fixes here:
> https://eggerapps-downloads.s3.eu-west-1.amazonaws.com/postico-5186.zip
> 
> We still need to do some further testing, but if everything works out we hope to release an update with these fixes soon.

The issue where by copying SQL from VSCODE (adding non-breaking spaces) was resolved for me in this build. Thx

**Comment by shmcsensei** [comment]
> > Hi everyone!
> > Sorry for the slow response to this bug report. I've further investigated the issue, and as far as I can tell, there are two separate issues:
> > 
> > 1. When pasting text from VS Code, all spaces are converted to non-breaking spaces
> > 2. Since PostgreSQL does not like non-breaking spaces, Postico tries to replace them with normal spaces (except in string literals), but the code that does this had a bug, and syntax highlighting didn't handle non-breaking spaces correctly either.
> > 
> > We've found a possible workaround for 1., and I've also attempted to finally fix the bugs from 2. I've also added a number of regression tests to make sure this problem doesn't happen again.
> > You can download the latest build with those fixes here:
> > https://eggerapps-downloads.s3.eu-west-1.amazonaws.com/postico-5186.zip
> > We still need to do some further testing, but if everything works out we hope to release an update with these fixes soon.
> 
> The issue whereby copying SQL from VSCODE (adding non-breaking spaces) was resolved for me in this build. Thx

That build although solved my issue seemed quite unstable. I caught it doing 100% and forced quit it after walking away from my computer for an hour. This is a deal breaker for me (I don't even know where to find older builds [not legacy]). I have worked at many companies and had them buy licences for this software. I really like postico (and paid) but until this is fixed I will unfortunately be using pgAdmin and telling people to do so likewise. Please fix this ASAP.

**Comment by jakob** [comment]
@shmcsensei 
I've reviewed recent changes and can't see anything new in build 5186 that would cause the CPU usage problem. It looks like you found a separate bug in Postico.

When the CPU usage problem happens again, could you create a sample of Postico and email it to me? Here's how to create a sample:

> To create a process sample, open Activity Monitor, select Postico, click the info button in the toolbar, then click "Sample". After a few seconds a sample will appear. Save it, then email it to us. Make sure to sample Postico while it is actually hanging.

My email address is jakob@eggerapps.at

With regard to your question about previous builds: Build 5186 was not yet released. This fix will be released in version 1.5.9. The current version is 1.5.8, which you can download from https://eggerapps.at/postico/

You can also download Postico 1.5.8 directly from this URL:
https://eggerapps-downloads.s3.eu-west-1.amazonaws.com/postico-1.5.8.zip
(Replace the version number in the URL if you want to download specific older versions.)

**Comment by jakob** [comment]
We just released Postico 1.5.9!

Thanks again to everyone for reporting this problem, and for your patience with this issue. Pasting from VS Code should now work as expected.
